programs/server/GRPCHandler.h

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+ #include <iostream>
+ #include <memory>
+#include <atomic>
+ #include <string>
+#include <mutex>
+ #include "IServer.h"
+ #include <Poco/RunnableAdapter.h>
+#include <IO/Progress.h>
+#include <Common/Stopwatch.h>
+ #include <IO/WriteBufferFromString.h>
+#include <DataStreams/BlockIO.h>
+#include <Processors/Formats/LazyOutputFormat.h>
+ #include <grpc++/server.h>
+ #include <grpc++/server_builder.h>
+ #include <grpc++/server_context.h>
+ #include "GrpcConnection.grpc.pb.h"
+
+ using GRPCConnection::HelloRequest;
+ using GRPCConnection::HelloResponse;
+ using GRPCConnection::QueryRequest;
+ using GRPCConnection::QueryResponse;
+ using GRPCConnection::GRPC;
+
+ namespace DB
+ {
+ class CommonCallData
+ {
+     public:
+     GRPC::AsyncService* Service;
+     grpc::ServerCompletionQueue* CompilationQueue;
+     grpc::ServerContext gRPCcontext;
+     
+     IServer* iServer;
+     bool with_stacktrace = false;
+     enum CallStatus { CREATE, PREPARE, PROCESS, ONFINISH, FINISH };
```

Comment:
These statuses make the code complicated and hard to understand.

Reply:
I looked at the function [RequestQuery](https://github.com/ClickHouse/ClickHouse/blob/2be1010cc0ab1db55674c19e6ec741d52ac41903/programs/server/GRPCHandler.cpp#L79). This function is declared in `GrpcConnection.grpc.pb.h` as follows
```
    void RequestQuery(::grpc::ServerContext* context, ::GRPCConnection::QueryRequest* request, ::grpc::ServerAsyncWriter< ::GRPCConnection::QueryResponse>* writer, ::grpc::CompletionQueue* new_call_cq, ::grpc::ServerCompletionQueue* notification_cq, void *tag) {
```
`new_call_cq` is responsible for finishing calls, and `notification_cq` is responsible for responding. You're passing the same completion queue for `new_call_cq` and for `notification_cq`. But you don't have to do so! Look:

```
GRPCServer(std::string server_address_, IServer& server_)
{
    ...
    notification_cq_ = builder.AddCompletionQueue();
    new_call_cq_.reset(new CompletionQueue());
    server = builder.BuildAndStart();
}

void HandleRpcs()
{
    new CallDataQuery(&service, new_call_cq.get(), notification_cq.get(), &iServer, log);
    
    auto handle_calls_completion = [&]()
    {
        while (true)
        {
            GPR_ASSERT(new_call_cq->Next(&tag, &ok));
            GPR_ASSERT(ok);
            delete static_cast<CallDataQuery*>(tag);
        }
    }

    auto handle_requests_completion = [&]
    {
        while (true)
        {
            GPR_ASSERT(notification_cq->Next(&tag, &ok));
            GPR_ASSERT(ok);
            auto thread = ThreadFromGlobalPool{&CallDataQuery::respond, static_cast<CallDataQuery*>(tag)};
            thread.detach();
        }

    notification_cq_thread = ThreadFromGlobalPool{handle_requests_completion};
    new_call_cq_thread = ThreadFromGlobalPool{handle_calls_completion};
}

CallDataQuery::CallDataQuery(GRPC::AsyncService* Service_, grpc::CompletionQueue* new_call_cq_, ServerCompletionQueue* notification_cq_, IServer* server_, Poco::Logger * log_)
{
Service->RequestQuery(&gRPCcontext, &request, &responder, new_call_cq, notification_cq, this);
}

CallDataQuery::respond()
{
    new CallDataQuery(Service, new_call_cq, notification_cq, iServer, log);

    context.setUser(request.user(), ...)
    ...
    parseQuery(...)
    ...
    executeQuery(...)
    ...
    responder.Write(...)
    ...
    responder.Finish()
}
```
No statuses at all!

Reply:
Or, if you're making the real streaming, then call `responder.Write()` from `WriteBufferFromGRPC::nextImpl()` and call `responder.Finish()` from `WriteBufferFromGRPC::finalize()`.